google_guest_agent/network/manager/systemd_networkd_linux.go

```diff
@@ -0,0 +1,236 @@
+	}
+	// The version is the second field of the first line.
+	versionString := strings.Split(strings.Split(res.StdOut, "\n")[0], " ")[1]
+	version, err := strconv.Atoi(versionString)
+	if err != nil {
+		return false, fmt.Errorf("error parsing systemd version: %v", err)
+	}
+	if version < minSupportedVersion {
+		logger.Infof("systemd-networkd version %v not supported: minimum %v required", version, minSupportedVersion)
+		return false, nil
+	}
+
+	// First check if the service is running.
+	res = run.WithOutput(ctx, "systemctl", "is-active", "systemd-networkd.service")
+	if res.ExitCode != 0 {
+		return false, nil
+	}
+
+	// Check systemd network configuration.
+	res = run.WithOutput(ctx, "/bin/sh", "-c", fmt.Sprintf("networkctl status %s --json=short", iface))
+	if res.ExitCode != 0 {
+		return false, fmt.Errorf("failed to check systemd-networkd network status: %v", res.StdErr)
+	}
+
+	interfaceStatus := make(map[string]any)
+
+	if err = json.Unmarshal([]byte(res.StdOut), &interfaceStatus); err != nil {
+		return false, fmt.Errorf("failed to unmarshal interface status: %v", err)
+	}
+
+	for _, statusKey := range n.networkCtlKeys {
+		state, found := interfaceStatus[statusKey]
+		if !found {
+			continue
+		}
+
+		return state == "configured", nil
+	}
+	return false, fmt.Errorf("could not determine interface state, one of %v was not present", n.networkCtlKeys)
+}
+
+// Setup sets up the non-primary network interfaces for systemd-networkd by writing
+// configuration files to the specified configuration directory.
+func (n systemdNetworkd) Setup(ctx context.Context, config *cfg.Sections, payload []metadata.NetworkInterfaces) error {
+	// Create a network configuration file with default configurations for each network interface.
+	googleInterfaces, googleIpv6Interfaces := interfaceListsIpv4Ipv6(payload)
+	for _, iface := range googleInterfaces {
+		logger.Debugf("write %s network config for %s", n.Name(), iface)
+
+		var dhcp = "ipv4"
+		if slices.Contains(googleIpv6Interfaces, iface) {
+			dhcp = "yes"
+		}
+
+		// Create and setup ini file.
+		data := systemdConfig{
+			GuestAgent: guestAgent{
+				ManagedByGuestAgent: true,
+			},
+			Match: systemdMatchConfig{
+				Name: iface,
+			},
+			Network: systemdNetworkConfig{
+				DHCP: dhcp,
+			},
+		}
+
+		config := ini.Empty()
+		if err := ini.ReflectFrom(config, &data); err != nil {
+			return fmt.Errorf("error creating config ini: %v", err)
+		}
+
+		// The number in beginning of the file name signifies its priority. We want this
+		// to take priority before any of the default configurations that systemd-networkd
+		// has set.
+		configPath := fmt.Sprintf("%s/50-%s-google-guest-agent.network", n.configDir, iface)
+		if err := config.SaveTo(configPath); err != nil {
+			return fmt.Errorf("error saving config for %s: %v", iface, err)
+		}
+
+		// Enable the interface without restarting systemd-networkd service.
+		if err := run.Quiet(ctx, "networkctl", "up", iface); err != nil {
```

Comment:
Only calling networkctl up will reload it, I'd assume it requires a down and up sequence.